src/replacer.rs

```diff
@@ -0,0 +1,334 @@
+            } else {
+                print!("{}", &line[current_index..*index]);
+            }
+            print!("{}", color(text));
+            current_index = index + text.len();
+        }
+        println!("{}", &line[current_index..]);
+    }
+}
+
+#[derive(Debug)]
+// A list of input_fragment, output_fragent
+struct Fragments(Vec<(Fragment, Fragment)>);
+
+impl Fragments {
+    fn new() -> Self {
+        Self(vec![])
+    }
+
+    fn is_empty(&self) -> bool {
+        self.0.is_empty()
+    }
+
+    fn add(&mut self, input: (usize, &str), output: (usize, &str)) {
+        let input_fragment = Fragment {
+            index: input.0,
+            text: input.1.to_string(),
+        };
+        let output_fragent = Fragment {
+            index: output.0,
+            text: output.1.to_string(),
+        };
+        self.0.push((input_fragment, output_fragent));
+    }
+}
+
+impl<'a> IntoIterator for &'a Fragments {
+    type Item = &'a (Fragment, Fragment);
+
+    type IntoIter = std::slice::Iter<'a, (Fragment, Fragment)>;
+
+    fn into_iter(self) -> Self::IntoIter {
+        self.0.iter()
+    }
+}
+
+/// Represent a fragment of text, similar to the data structure returned
+/// by String::match_indices
+#[derive(Debug)]
+struct Fragment {
+    index: usize,
+    text: String,
+}
+
+trait Replacer {
+    // return position of the match, input_text, output_text, or None
+    fn replace(&self, buff: &str) -> Option<(usize, String, String)>;
+}
+
+struct SubstringReplacer<'a> {
+    pattern: &'a str,
+    replacement: &'a str,
+}
+
+impl<'a> SubstringReplacer<'a> {
+    fn new(pattern: &'a str, replacement: &'a str) -> Self {
+        Self {
+            pattern,
+            replacement,
+        }
+    }
+}
+
+impl<'a> Replacer for SubstringReplacer<'a> {
+    fn replace(&self, buff: &str) -> Option<(usize, String, String)> {
+        let pos = buff.find(&self.pattern)?;
+        Some((pos, self.pattern.to_string(), self.replacement.to_string()))
+    }
+}
+
+struct SubvertReplacer<'a> {
+    items: &'a [(String, String)],
+}
+
+impl<'a> SubvertReplacer<'a> {
+    fn new(items: &'a [(String, String)]) -> Self {
+        Self { items }
+    }
+}
+
+impl<'a> Replacer for SubvertReplacer<'a> {
+    fn replace(&self, buff: &str) -> Option<(usize, String, String)> {
+        // We need to return the best possible match, otherwise we may
+        // replace FooBar with SpamEggs *before* replacing foo-bar with spam-eggs
+        let mut best_pos = buff.len();
+        let mut best_index = None;
+        for (i, (pattern, _)) in self.items.iter().enumerate() {
+            let pos = buff.find(pattern);
+            match pos {
+                Some(p) if p < best_pos => {
+                    best_pos = p;
+                    best_index = Some(i)
+                }
+                // Not found, or in a later position than the best, ignore
+                _ => {}
+            }
```

Comment:
The match can be an `if`, since it only has one arm
```
            if let Some(pos) = buff.find(pattern) {
                if pos < best_pos {
                    best_pos = pos;
                    best_index = Some(i)
                }
            }
```